test.sh

```
@@ -281,8 +281,24 @@ run_pytest() {
+    local basename_target
+    basename_target="$(basename "$target")"
+    local source="/coverage/$(basename "$target")"
+
+    # E2E exercises API/worker/scheduler in sibling containers. Stop those
+    # processes before materializing the report so coverage.py flushes their
+    # parallel data files into the shared coverage volume.
+    docker compose -f "$COMPOSE_FILE" stop api worker scheduler >/dev/null 2>&1 || true
+
     docker compose -f "$COMPOSE_FILE" --profile test run --rm test-runner \
-        sh -lc 'cat /coverage/coverage.xml' > "$target"
+        sh -lc "backup=/tmp/$(basename '$source').pre-combine
```

Shellcheck (0.11.0)

[info] Expressions don't expand in single quotes, use double quotes for that.

(SC2016)